### PR TITLE
Milestone3 - Game logic for Tic-Tac-Toe game in Ruby

### PR DESCRIPTION
**Pull request by ybadmus**
This is the third project of Ruby, It's an introduction to the Object Oriented Concepts of the Ruby Programming Language. We built a Tic-Tac-Toe game using several concepts of OOP to understand the syntax for achieving them in Ruby.

### CODE REVIEW

**Review by ebeagusamuel — CHANGES_REQUESTED** [review]
# Changes Required :recycle: 
### Hey there :wave: 
###### Nice attempt at writing the logic for the game and trying to get the game to work as expected. You have done well so far but there is still some work to be done to get the game to work just fine and meet the requirements for the project to be approved.
- [ ] All game-logic related resources (classes and/or modules) are kept in the lib directory.
- [ ] The application can determine if the move is a winning move
  * [ ] The application can detect a sequence of three horizontal elements that are equal
  * [ ] The application can detect a sequence of three vertical elements that are equal
  * [ ] The application can detect a sequence of three equal elements in a diagonal from left and right to right
- [ ] The application can tell if there was a draw after players move
#### Currently your application uses the same symbol for both players, which shouldn't be so. Also, when a player makes an invalid move, your application informs the players of the invalid move, and instead of it to ask the player that made the invalid move to make another move, it asks the other player. 
###### Please make changes accordingly and request another review when you are done.
slack(@ebeagusamuel)

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
# Status: Changes Required :vertical_traffic_light: :warning: 
## Hello @ybadmus  :handshake: :handshake: :handshake: 

:+1: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.
:+1: I liked that you have tried to implemented the requested changes from the previous review, keep it up, and your project is really almost done.
#### :warning: However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below :point_down: :point_down: :point_down: 
### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes] :red_circle: 
- [ ] Currently your game cannot detect these winning moves: 
  - [ ] All diagonals
  - [ ] Middle horizontal
=> please fix it so that it detect them all
___
_If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._

[GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

_Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

**Comment by ybadmus** [comment]
@descholar-ceo I will be glad if you can give examples of what you are talking about, I just cross checked everything again and all the diagonals are passing. The Middle horizontal is also passing the test, I have attached a screenshot. Kindly check again and approve, I really want to move to the next step. Thanks
middle-diagonal

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required 🛑

## Hello team 👋,

**Great job implementing the requested changes 👏. However, there are a few requirements needed.**

1). **The middle horizontal win is not being detected. Check the screenshot below.**
win

2). **The middle and last vertical wins are not being detected. Check the screenshot below.**
win3

3). **The move below should be a win for the player, yet it says it is a tie.**
win2


## [Optional]

- You can add validations to ensure a username is not blank.
- Display the gameboard even after the game is won.

## Happy coding 🥂.

**You can reach me on slack (Rose Wanjohi).**

File: bin/main.rb [hunk 1]
```diff
@@ -1,89 +1,46 @@
...
-# rubocop:disable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
 
-class Users
-  MAX_COUNTS = 9
-  SUCCESS_OUTCOMES = [
-    [1, 2, 3], [4, 5, 6], [7, 8, 9], [1, 4, 7],
-    [2, 5, 8], [3, 6, 9], [1, 5, 9], [3, 5, 7]
-  ].freeze
+require './lib/users'
+require './lib/logic'
```
Comment (rindrajosia) [inline bin/main.rb]: Use **require_relative** for files within the folder structure. **Require** should be used for external files, like gems, while **require_relative** should be used for referring to files within your directory.
In addition to that, check the **path** here, it should be **'../lib/users**' and '**../lib/logic**'.

File: lib/logic.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  def available_moves
+    puts <<-HEREDOC
+      ---+---+---
+      #{@cells[0]} | #{@cells[1]} | #{@cells[2]}
+      ---+---+---
+      #{@cells[3]} | #{@cells[4]} | #{@cells[5]}
+      ---+---+---
+      #{@cells[6]} | #{@cells[7]} | #{@cells[8]}
+      ---+---+---
+    HEREDOC
```
Comment (rindrajosia) [inline lib/logic.rb]: No **puts** outside **main.rb**

**Review by rindrajosia — CHANGES_REQUESTED** [review]
**Changes Required**
Hi!!
Very good job! Overall, this project is awesome! You are VERY close to finishing it! 

We only require to change some minor issues in order to get your project approved:
- [ ] [ REQUIRED] No **puts** outside **main.rb** file ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
- [ ] [ REQUIRED] Use **require_relative** for _files within the folder structure_

After addressing those minor bugs submit another CRR.

**Happy Coding**! 
Rindra Josia

**Comment by mimipeshy** [comment]
### STATUS :APPROVED :partying_face: 

Hey there @ybadmus  :wave: :wave:  good job on working on the requested changes. This project is therefore approved. Go ahead and merge. All the best in the next milestones :man_technologist: 

Regards
Peris Ndanu

**Comment by mimipeshy** [comment]
### STATUS :APPROVED :partying_face: 

Hey there @ybadmus  :wave: :wave:  good job on working on the requested changes. This project is therefore approved. Go ahead and merge. All the best in the next milestones :man_technologist: 

Regards
Peris Ndanu

**Pull request merged**